[Cocoa] Adopt -makeStreamingContentKeyRequestDataForApp in CDMSessionAVContentKeySession

jernoble opened the pull request:
#### b4568cba88d9e148145c207ab61c9daacbe3cf16
<pre>
[Cocoa] Adopt -makeStreamingContentKeyRequestDataForApp in CDMSessionAVContentKeySession
<a href="https://bugs.webkit.org/show_bug.cgi?id=266264">https://bugs.webkit.org/show_bug.cgi?id=266264</a>
<a href="https://rdar.apple.com/119527921">rdar://119527921</a>

Reviewed by Eric Carlson.

Adopt -makeStreamingContentKeyRequestDataForApp, which takes a completion handler that is called on
a background queue, but force the call to become synchronous through a local semaphore.

* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionAVContentKeySession.mm:
(WebCore::CDMSessionAVContentKeySession::update):

Canonical link: <a href="https://commits.webkit.org/271933@main">https://commits.webkit.org/271933@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/507c401e63d97ccec366b0d07076fd1a642923b6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/30039 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8701 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/31349 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32537 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/27150 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/30699 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10903 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5951 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/27180 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/30333 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/7303 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/25610 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/6224 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/6391 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/26702 "Found 1 new test failure: imported/w3c/web-platform-tests/content-security-policy/reporting-api/reporting-api-works-on-frame-src.https.sub.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33872 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/27398 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/27135 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/32562 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/6326 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4500 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/30360 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/8075 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7130 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/7078 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6853 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionAVContentKeySession.mm:
```diff
@@ -313,14 +313,22 @@ static bool isEqual(Uint8Array* data, const char* literal)
+            requestData = outRequestData;
+            keyRequestCompleted.signal();
+        }];
+        keyRequestCompleted.wait();
```
Should this use `waitFor` so we can timeout, like we do when calling `processContentKeyRequestWithIdentifier`?

jernoble replied:
I intentionally did not do that, because the block is referencing stack variables. If we timeout and continue, I suspect that the block will run after the stack has been torn down and cause a UAF.

jernoble committed:
[Cocoa] Adopt -makeStreamingContentKeyRequestDataForApp in CDMSessionAVContentKeySession
https://bugs.webkit.org/show_bug.cgi?id=266264
rdar://119527921

Reviewed by Eric Carlson.

Adopt -makeStreamingContentKeyRequestDataForApp, which takes a completion handler that is called on
a background queue, but force the call to become synchronous through a local semaphore.

* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionAVContentKeySession.mm:
(WebCore::CDMSessionAVContentKeySession::update):

Canonical link: https://commits.webkit.org/271933@main
